File: src/lib/types/app-state.ts
/**
 * Application state machine contract.
 *
 * The Rust backend owns the authoritative state machine (`bluey_core::state`).
 * The frontend mirrors it via the `app.state` event and never invents its own
 * `isLoading` / `isThinking` booleans. `audioActive` is the single orthogonal
 * region: an audio session can be running while a capture/think cycle happens.
 */

import type { BlueyError } from "./errors";

export type AppState =
  | "booting"
  | "auth_required"
  | "ready"
  | "listening"
  | "capturing"
  | "analyzing"
  | "thinking"
  | "response_ready"
  | "error"
  | "paused";

export interface AppStatus {
  state: AppState;
  /** True while an audio session is running (orthogonal to the primary state). */
  audioActive: boolean;
  /** Active session id, if a session is running. */
  sessionId?: string;
  /** Active mode id (global default or session override). */
  modeId: string;
  /** Present when `state === "error"`. */
  error?: BlueyError;
  /** State we will return to after `error`/`paused` is resolved. */
  resumeState?: AppState;
  /** ISO timestamp of the last transition. */
  updatedAt: string;
}

/** Events accepted by the state machine (mirrors `bluey_core::state::AppEvent`). */
export type AppStateEvent =
  | { type: "boot_completed"; authenticated: boolean }
  | { type: "authenticated" }
  | { type: "signed_out" }
  | { type: "audio_started" }
  | { type: "audio_stopped" }
  | { type: "capture_started" }
  | { type: "capture_finished" }
  | { type: "analysis_started" }
  | { type: "thinking_started" }
  | { type: "response_ready" }
  | { type: "response_dismissed" }
  | { type: "failed"; error: BlueyError }
  | { type: "recovered" }
  | { type: "paused" }
  | { type: "resumed" }
  | { type: "session_changed"; sessionId?: string }
  | { type: "mode_changed"; modeId: string };

export const IDLE_STATES: ReadonlySet<AppState> = new Set<AppState>(["ready", "listening"]);
export const BUSY_STATES: ReadonlySet<AppState> = new Set<AppState>([
  "capturing",
  "analyzing",
  "thinking",
]);

File: src-tauri/crates/bluey-core/src/types/app_state.rs
use serde::{Deserialize, Serialize};

use crate::error::BlueyError;

/// Mirrors `AppState` in `src/lib/types/app-state.ts`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AppState {
    Booting,
    AuthRequired,
    Ready,
    Listening,
    Capturing,
    Analyzing,
    Thinking,
    ResponseReady,
    Error,
    Paused,
}

impl AppState {
    pub fn is_idle(&self) -> bool {
        matches!(self, AppState::Ready | AppState::Listening)
    }
    pub fn is_busy(&self) -> bool {
        matches!(
            self,
            AppState::Capturing | AppState::Analyzing | AppState::Thinking
        )
    }
}

/// Mirrors `AppStatus`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppStatus {
    pub state: AppState,
    pub audio_active: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    pub mode_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<BlueyError>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub resume_state: Option<AppState>,
    pub updated_at: String,
}

/// Mirrors `AppStateEvent` (tag = "type").
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum AppEvent {
    BootCompleted {
        authenticated: bool,
    },
    Authenticated,
    SignedOut,
    AudioStarted,
    AudioStopped,
    CaptureStarted,
    CaptureFinished,
    AnalysisStarted,
    ThinkingStarted,
    ResponseReady,
    ResponseDismissed,
    Failed {
        error: BlueyError,
    },
    Recovered,
    Paused,
    Resumed,
    #[serde(rename_all = "camelCase")]
    SessionChanged {
        #[serde(default, skip_serializing_if = "Option::is_none")]
        session_id: Option<String>,
    },
    #[serde(rename_all = "camelCase")]
    ModeChanged {
        mode_id: String,
    },
}
